dali/python/nvidia/dali/_debug_mode.py

```diff
@@ -0,0 +1,332 @@
+        self._es_kwarg_name = 'kwarg_'
+        self._subpipeline_kwargs = kwargs
+        self._subpipelines = {}
+        self._external_source_pipelines = {}
+        self._feed_input_data = {}
+        self._subpipelines_built = False
+        self._cur_subpipeline_id = -1
+        self._exec_func = exec_func
+        self._runner = _PipelineDebugRunner(self)
+
+    def __enter__(self):
+        raise RuntimeError("Currently pipeline in debug mode works only with `pipeline_def` decorator."
+                           "Using `with` statement is not supported.")
+
+    def build(self):
+        """Build the pipeline.
+        
+        Refer to :meth:`Pipeline.build() <nvidia.dali.Pipeline.build>` for details."""
+        self._built = True
+
+    def run(self):
+        """Run the pipeline and return the result."""
+        import numpy as np
+        if not self._built:
+            raise RuntimeError('Pipeline must be built first.')
+        with self._runner:
+            res = self._exec_func()
+            if res is None:
+                res = ()
+            elif not isinstance(res, tuple):
+                res = (res,)
+
+            # Transforming all variables to TensorLists.
+            return tuple([val.get() if isinstance(val, DataNodeDebug) else _Tensors.TensorListCPU(
+                np.tile(val, (self._max_batch_size, *[1]*np.array(val).ndim))) for val in res])
+
+    def feed_input(self, data_node, data, **kwargs):
+        """Pass data to an ExternalSource operator inside the pipeline.
+        
+        Refer to :meth:`Pipeline.feed_input() <nvidia.dali.Pipeline.feed_input>` for details."""
+        if not self._built:
+            raise RuntimeError("Pipeline must be built first.")
+
+        if data_node not in self._external_source_pipelines:
+            # Saving data, because pipeline hasn't been run yet.
+            if data_node not in self._feed_input_data:
+                self._feed_input_data[data_node] = []
+            self._feed_input_data[data_node].append((data, kwargs))
+        else:
+            self._external_source_pipelines[data_node].feed_input(data_node, data, **kwargs)
+
+    @staticmethod
+    def _classify_data(data):
+        # Returns tuple (use_external_source, device, unpacked data).
+
+        def is_primitive_type(x):
+            return isinstance(x, (int, float, bool, str))
+
+        if isinstance(data, list):
+            if any([is_primitive_type(d) for d in data]):
+                return False, 'cpu', data
+
+            device = None
+            data_list = []
+            for d in data:
+                _, cur_device, val = _PipelineDebug._classify_data(d)
+                if device == None:
+                    device = cur_device
+                data_list.append(val)
+            return True, device, data_list
+        else:
+            if is_primitive_type(data) or _types._is_numpy_array(data) or _types._is_mxnet_array(data) or isinstance(data, _Tensors.TensorCPU):
```

Comment:
I think we can also have mxnet ndarray that is on gpu.

Reply:
I took [https://docs.nvidia.com/deeplearning/dali/main-user-guide/docs/supported_ops.html?highlight=external_source#nvidia.dali.fn.external_source](url) as a point of reference.

Reply:
Sure, let's leave it for now.